KEP: Pod Schedule Readiness (#3522)
kubernetes/enhancements: keps/sig-scheduling/3521-pod-schedule-readiness/README.md

Integrators can mutate this field to instruct scheduler whether it's ready for scheduling or not.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

Pods are naturally perceived ready for scheduling once created. Kubernetes scheduler does
its due diligence to find nodes to place all pending Pods. However, in a real-world case,
some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
unnecessary manner.

Lacking the knob to flag Pods as scheduling-ready/unready wastes scheduling cycles on retrying
Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to

> develop some in-house features (such as hierarchical quota) natively.

Reviewer: Agree with this point. Important to see this defined. The pods that cannot be scheduled as they are paused must not trigger autoscaling.

> develop some in-house features (such as hierarchical quota) natively.

Author: I listed it as a non-goal, but not quite descriptive I guess, going to reword it a bit:

```
### Non-Goals

- Mark scheduling-unready Pods as `Unschedulable` by updating their `PodReady` condition.
- ...
```

> develop some in-house features (such as hierarchical quota) natively.

Reviewer 2: a non-goal can become a goal in the future, and that should never be the case. So it should be list as a goal to not mark them as unschedulable.

> develop some in-house features (such as hierarchical quota) natively.

Author: updated.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

- Scheduler doesn't actively clear a Pod's `schedulingPaused` state. This means if the controller
logic is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
in Pending state for a long time and it carries `schedulingPaused=true`, you may find out which component
updates the state (via `.spec.managedFields`) and report the symptom to the component owner.

- Faulty controllers may forget to unset the Pod's `schedulingPaused` state, and hence results in
a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulabe Pods caused
by potential faulty controllers. We will evaluate necessary options in the future to mitigate
potential abuse.

- End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We

> will provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.

Reviewer 2: creating an event would be appropriate

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API

A new API field `schedulingPaused` will be added to Pod's spec:

```go
type PodSpec struct {
    <<[UNRESOLVED]>>
    // Each scheduling gate corresponding with the condition name of .status.conditions.
    // Gating condition can be updated by different controller, and scheduling is triggered
    // only when all gating conditions are ready/true.
    // This option will need to be accompanied with a yet-to-be-implemented fined-grained
    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
    // and consistent with how finalizes/liens work today.

>     SchedulingGates []string

Reviewer 2: @smarterclayton it seems that now you are leaning towards this option vs a flag, correct? If so, we also need to discuss the format, we now have two:

1) similar to finalizers: actors add to the list and later remove from the same list
2) similar to readinessGates: actors add to the list in the spec, and later add to a parallel list in status

Regarding option 2, do we have concerns that it will require controllers to have update permissions on pod status? I guess not if we allow updating the list only through a dedicated endpoint since we want to enable fine grained permissions.

I like option 2 because it preserves the history of who decided that a pod is ready, which is useful for debugging and tracking feature usage.

Author: Personally, I'm in favor of `readinessGates` as (a) we have pod readiness gates as precedence, and (b) each controller don't need to worry about updating the same API field, and thus keep some factors like ordering away. However, two aspects are still unresolved (https://github.com/kubernetes/enhancements/pull/3522#discussion_r976723138)

- Permission. It's still unclear to me. The [pod readiness gates](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/580-pod-readiness-gates/README.md) KEP didn't mention about permission. So I really hope @smarterclayton and @liggitt can shed some light - as if we think permission is mandatory here, we have to make it right since day 1.
- Restrictive condition transition. I suppose it's doable to enforce the condition that is transited from NotReady to Ready only. But reposted here for confirmation.

Reviewer 3: I need to read the arguments around pod readiness gates and the whole status thing.  Scheduling happens before ownership is transitioned to a node, I would probably argue that like finalizers, all mutations happen within spec (i view metadata as generic spec in this scenario).

Anyone who can set spec.nodeName is already as powerful as someone who can set the proposed scheduling gates.  OpenShift and some others had admission checks that limited who could mutate that field to effectively cluster admins (roughly the [mutation of that field required an RBAC check](https://github.com/openshift/kubernetes/blob/master/openshift-kube-apiserver/admission/scheduler/podnodeconstraints/admission.go#L135) you have to have the binding subresource permission of a scheduler).

Subresource permission is supported today, so adding a subresource is the easiest way, but we'd probably have to have special cases for what characteristic would be permission (i.e. a check on `readiness/<name>`), but it's not terribly difficult today to make the sub resource `pods/readiness/<name>`.  The more we start bulk updating gates (for instance if the scheduler implements a set of gates via plugins and we want to bulk submit, vs letting individual actors coordinate) the more we get closer to just performing field level checks in an admission handler.  

How would we model multiple gates coordinating in a single scheduler?  I would very much dislike having multiple individual writes to the object, but anything that happens in a single process and is fairly quick should probably just be a single gate (even if it's implemented as a composite of steps). 

Ultimately subresources work when the use case is narrow and does not composite across fields - this is starting to cross that line for binding if we expect lots of gates from a single client being set in a row.

Reviewer 2: > The more we start bulk updating gates (for instance if the scheduler implements a set of gates via plugins and we want to bulk submit, vs letting individual actors coordinate) the more we get closer to just performing field level checks in an admission handler.

I don't think the scheduler itself will have plugins that update gates, the scheduler will only respect those gates (i.e., if they are present, it will simply not attempt to schedule the pod). 

The gates will be updated by external controllers, like a queueing controller or a statefulset migration controller, those by design will likely be controllers that don't share memory and run as separate processes. I think the proper design pattern should be that a controller only handles/updates a single gate.

Author: No matter if the field is a single boolean or a list of strings, a Pod's creation is the _only_ chance to set the **complete** gating condition(s). For a single boolean, `true` represents the condition; while for a list of strings, it's highly customized list. After a Pod's creation, only "semantical subtraction" can be allowed: for boolean, it's true -> false; for a list of strings, it's "delete a particular element from the list" (it just behaves differently for finalizer-like list vs. readiness gates-like list)

> I would very much dislike having multiple individual writes to the object

For the design of a list of strings, both options (finalizer-like and readiness gate-like) will inevictably yield N individual writes to a Pod - each controller needs to spawn an API call to relax the readiness gate anyway (either by deleting one entry from the finalizer-like lists, or updating their particular .status.conditions[x] to ready/true).

Reviewer 3: Sorry, I should have said "N writes by the same client".  "N writes by N clients" is fine.  "2N writes by N clients"... might be ok, but I want to ask here.  What sort of gates would we imagine?  Someone filling out scheduling constraints (fleshing out topologyConstraints)?  Adding tolerations?  Adding resources?  The more general purpose, the more I want to at least consider avoiding `2N` writes in the design of the mechanism.  The more general purpose, the more we want to avoid `MN` RBAC checks on fields.  I like general purpose mutation control, but I hate to overly couple a new design like this with it, as we noted in the subresource doc.

Just to be 100% certain, we don't think we'd have an internal to the scheduler gate?

Author: > What sort of gates would we imagine?

The gates can be related to in-house multi-tenancy (e.g., advanced quota control), batch support (e.g., gang-scheduling so start scheduling only when a quorum is satisfied), etc.

> I want to at least consider avoiding 2N writes in the design of the mechanism

I'm actually more aggressive to restrict it a N writes at maximum. That's to say, each gate owner has only one chance to turn the flag from paused to unpaused.

> Just to be 100% certain, we don't think we'd have an internal to the scheduler gate?

Do you mean if scheduler would be manipulating the gates? No, scheduler would not mutate a Pod's spec in any way. The only fields it update is status.condition, status.nominatedNodeName and v1.binding (.spec.nodeName).

>> a Pod's creation is the only chance to set the complete gating condition(s). 
>
> I didn't anticipate this being required. Gate 1 could add 2 more gates and remove itself. That was something we believed we should support in initializers, because it allowed chaining at the cost of more stages. Is there a reason not to allow it?

That makes the scheduling racy. Suppose all gates are lifted, and the scheduler to schedule a Pod, and meanwhile then new gate(s) comes in aiming to pause the Pod, but due to current scheduling design, that "gating" signal cannot stop an in-flight pod's scheduling. 

This is also consistent what [Pod Readiness Gate](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/580-pod-readiness-gates#implementation-detailsnotesconstraints): "ReadinessGates can only be specified at pod creation."

> Do we have a use case that requires gates to be processed in order, vs being parallel?

For now I'm not aware of any req on this. Say, we assume each gate's logic is independent.

Reviewer 3: > That makes the scheduling racy. Suppose all gates are lifted, and the scheduler to schedule a Pod, and meanwhile then new gate(s) comes in aiming to pause the Pod, but due to current scheduling design, that "gating" signal cannot stop an in-flight pod's scheduling.

Sorry, I meant "the success of gate 1 is indicated at the same time gate 2 and 3 are added".  That is not racy because the scheduler observes version 1 with "gate1", then version 2 with "gate1 (complete), gate 2, gate 3".  But you're right, for us to deviate from readiness gates consistency we'd need a use case justifying it.

Even though I don't like mutating webhooks injecting things on pod creation (every additional webhook reduces total cluster reliability because we've added a new independent source of failure), an approach where a third party used mutable gates to behave like initializers (instead of doing it at creation time, you do it via a regular controller) changes it from a problem of the api client (get a fail closed on the webhook) to a problem visible to the scheduler (the gate initializer controller hangs).  That said, if we're going to have CEL eventually, you could use CEL to set up your basic gates and let a more sophisticated gate implementation (the batch scheduler) handle the details, including potential delegation.

I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely?  I think so, in which case we can start with the more limited option and see how we go.

Reviewer 2: I agree that we don't need to enforce order, the base use case is a group of independent controllers that don't really coordinate.

> I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely? I think so, in which case we can start with the more limited option and see how we go.

+1, I think it is safe to start strict and potentially relax later, especially if this field will be gated via RBAC.

At this point I think we have two issues to resolve:
1. API format: finalizer vs readinessGate format. How can we move forward from here? 
2. How to gate this field, subresource can be mentioned as an option for now, but we could also say that we will wait for a recommendation from sig-auth.

Author: > "the success of gate 1 is indicated at the same time gate 2 and 3 are added"

If the 3 operations are completed within one API call, then yes it's not racy, but as you mentioned, we may start with the most strict version as a start.

> I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely? I think so, in which case we can start with the more limited option and see how we go.

+1 as well.

> 1. API format: finalizer vs readinessGate format. How can we move forward from here?

For readinessGates, to confirm we're on the same page that:

- a) each gate corresponds to a condition with the same name as the condition's `type`
- b) specifying gates as Pod's creation time indicates the corresponding conditions are "paused" initially
- c) nil conditions{gates[x]} is treated as a legal "paused" condition
- d) there will be validation logic enforcing conditions{gates[x]} is updated in an one-way ("paused"->"unpaused") direction
- e) in Alpha, `gates` are immutable after creation time

If we all agree on this, I'd be glad to update the doc to turn a single boolean to this.

> 2\. How to gate this field, subresource can be mentioned as an option for now, but we could also say that we will wait for a recommendation from sig-auth.

A practical way is to document the auth restriction in Alpha, and revisit it in Beta (hope we will have a canonical auth way by then). WDYT?

Author: Nope, we don't expect individual end-user to patch the status. Only privileged/trusted controllers can patch.

Author: Yes, correct. Regarding Pods that are not subject to schedulingGates, they are scheduled as before.

Reviewer 2: One potential issue is if by mistake the same gate name was used for both pod and scheduling readiness. Perhaps we should validate that readinessGates and schedulingReadinessGates are mutually exclusive.

Author: yes, currently every readiness gate needs to be qualified, so no harm to add additional validation checks on the uniqueness of gate names.

Author: @smarterclayton as well. We may tend to adopt schedulingGates over finalizer-like string list.

Reviewer 3: Re: string list vs array list - it is strictly preferable in almost all cases on a single type to bias away from `[]primitive` to `[]struct` - there is almost no downside because the field can be extensible and the serialization isn't substanitally more complex.  

For finalizers, the argument came down to whether a field in metadata would be extensible - we could not envision it (it's really hard to extend finalizer generically) and we had annotations and labels to carry the extensibility for custom finalizers if we needed it (or per type fields).  So I would argue finalizers are special because they're in metadata, while every other type should prefer `[]struct`.

I'm ok with permission bits coming after alpha, it's no worse than finalizers, leins, spec.nodeName, or a few other fields today.

Re: mutually exclusive - just to verify, is the problem that the gate names go into the same list somewhere?  I don't see how that's possible, readiness is a different part of the lifecycle than scheduling (by definition, a pod that hasn't been scheduled can't be ready)?

Author: > Re: string list vs array list

The only concern is it's not that consistent with the API design of current readiness gates. Readiness gates are of type []string, and each gate corresponds to status.conditions[x]. But personally I have no strong option on []string over []struct. Given readiness gates have been a while, not quite its API pattern is still a best practice. @liggitt @ahg-g 

> just to verify, is the problem that the gate names go into the same list somewhere?

It's based on using `[]string` as the schedulingGates, so the mapped condition needs to be unique - each gate name (scheduling or regular for network) corresponds to a condition with its `type` as the same name (required to be a qualified naming style: foo.com/xyz).

Reviewer 2: @smarterclayton I don't think you explicitly made your recommendation regarding finalizers vs readinessGates format. The former is a list that we remove from, the latter is a static list of gates and PodConditions are used to "open" the gates.

> Re: mutually exclusive - just to verify, is the problem that the gate names go into the same list somewhere? I don't see how that's possible, readiness is a different part of the lifecycle than scheduling (by definition, a pod that hasn't been scheduled can't be ready)?

Mutually exclusive is relevant if we go with the podReadiness format. The lists where we declare the gates are separate in the pod Spec, but both features will use PodConditions to "open" the gates. Think of the case where the same gate is declared in both lists by mistake, for example two unrelated podReadiness and schedulingReadiness controllers using the same gate name.

Reviewer 3: > The former is a list that we remove from, the latter is a static list of gates and PodConditions are used to "open" the gates.

Sorry, I was implying that pod conditions weren't the right approach (no status, so no conditions), so implication being that the modelling would be a `spec.[]struct` that clients remove from OR simply set a field (like finalizers).  I was emphasizing the spec structure of readinessGates vs the "spec + status coordination around conditions".  Conditions are appropriate for multiple clients to coordinate on once the pod passes to node ownership (spec.nodeName is set), but status is not appropriate for multiple clients to coordinate on prior to the pod being scheduled (IMO, open to arguments).

ReadinessGates can be mutable over time, SchedulingGates are one way.  Whether we remove or flag the spec with some status, we should consider whether we need a record on the pod of the gate being passed.  I would probably say we don't need a record (we didn't need a record for finalizers).

Author: Actually I agreed with the interpretation of .status.conditions, and that's why I mentioned a couple of time the concern of "enforcing one-way transition on conditions[x]". Enforcing on conditions doesn't quite have a precedence, although in readiness gates's KEP, it mentioned the expectation:

> To conform with this proposals, workload controllers MUST take pod “Ready” condition as the final signal to proceed during transitions

But I checked the code, there is no enforcement logic.

Reviewer 4: > Conditions are appropriate for multiple clients to coordinate on once the pod passes to node ownership (spec.nodeName is set), but status is not appropriate for multiple clients to coordinate on prior to the pod being scheduled (IMO, open to arguments).

I'm not sure I buy the pre/post-node-assignment distinction on using pod.status / pod.status.conditions

Wanting clarity on the implications of building things on top of conditions that expect/require one-way semantics is fair

Author: > Can you describe what code you checked? You mean, readinessGate has no enforcement of one way flip?

I checked the following code paths:

- pkg/apis/core/validation/validation.go#validatePodConditions
- pkg/apis/core/validation/validation.go#validateReadinessGates

Also, I tested it on a kind v1.25 env: https://gist.github.com/Huang-Wei/734b468571d8bd3a8e3211f34c197414

Reviewer 3: > I'm not sure I buy the pre/post-node-assignment distinction on using pod.status / pod.status.conditions

Fair, I think it's more "conditions as parts of a state machine feel wrong" vs "pre / post kubelet ownership".

Conditions don't feel like they are the appropriate field to represent a state machine, especially since most gates would be unlikely to need to convey information to a user.  Whereas a failure to go ready in a pod is a more granular operation that users are likely to debug.

Your one-way point is clearer - nothing stops an integrator from adding a condition to indicate failure to proceed, but the semantics of conditions do not include immutability today.  Whereas we can add such semantics to a new field fairly cleanly.  Hopefully per condition type immutability semantics is not seen as desirable for a consumer today (certainly not practical to add in the near term).

As an aside, i'm now worried about readiness gates not taking into account that node state can be lost without the pod being deleted - a pod that has been ready once may have gates on the node that need to be reset in some pod recovery scenarios.  I hope we don't have gates making such an assumption and being incorrect on failure...

> I checked the following code paths

Yeah, someone adding one way immutability to conditions would be questionable - conditions don't have those semantics today, this seems like tightening validation.

Reviewer 2: It looks like now we have a stronger reason not to pursue readinessGates semantics. Although keeping a record of the gates would be really nice for debug-ability, there are obviously other ways we can achieve that (logging and events at the time of pod creation).

> That might be "no other finalizers before me in the list", that could also be coordination with other objects. We should as part of this design consider some normative / recommended coordination documentation similar to finalizers (since finalizers are a stronger phase transition construct than readiness gates).

I guess the only recommended coordination we can document is that dependency between gate managers should be one way to avoid circular dependencies.

Author: Re: "conditions as parts of a state machine feel wrong"

Can we bypass the enforcement on conditions just like readiness gates? and also document it that the integrator should be fully responsible for it. In such cases, we have the freedom to reset the condition in some corner cases like Clayton mentioned.

Reviewer 3: Hrm.  In essence, we have a spectrum:

1. Initializers / finalizers - `[]string`, removal only, additional data stored on object or at worst in meta, extremely unlikely we add supporting data
2. schedulingGates - removal only, additional data stored on pod spec, potentially other data in the future but unlkely, users unlikely to need the status of each gate, integrators likely able to edit spec
3. readinessGates - `[]struct + conditions`, readiness is inherently reactionary and two-way, gates say they are one way but are not (and probably can't safely change conditions), it think it is possible to lose all node state in such a way that gates have to be re-executed (but we don't clearly define that anywhere), integrators SHOULD NOT be able to edit spec, users want to see the status of each gate

I think the argument I'm thinking about is that `nothing requires conditions, therefore why are they a part of the flow`.  Consistency is one argument, but I think the use case is closer to initializers / finalizers, and since we don't have a *need*, I'd argue YAGNI.  If we do get feedback that users want a lot of condition state from gates, we could reassess in beta (the structure wouldn't change for our spec).

Thoughts?

Reviewer 2: I agree with option 2, and I think even now integrators are free to add conditions to the pod if they wish to offer more status information, don't they? So there is no need to tie that to our feature.

Reviewer 3: Right, good argument.

Author: Option 2 should work for most cases. If there is some case like re-scheduling a pod without deleting it, it's also extensible / backward-compatible to add fields.

Author: updated.

- Benchmark tests passed, and there is no performance degradation.
- Update documents to reflect the changes.

#### GA

- Fix all reported bugs.
- Feature enabled and cannot be disabled. All feature gate guarded logic are removed.
- Update documents to reflect the changes.

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

- Upgrade
  - While the feature gate is enabled, `.spec.schedulingPaused` can be mutated by users.
  - While the feature gate is enabled and `.spec.schedulingPaused` is not set, a default value will
  be given to behave the same like it's specified by users.
- Downgrade
  - Previously configured values will be ignored.

> 

Reviewer 2: We need to mentioned that this is an opt-in feature since users need to explicitly set the field in the pod spec. If the feature is disabled the field is preserved if it was already set in the persisted pod object, otherwise it is silently dropped.

Author: I suppose the downgrade means "downgrading to a version that this feature is not supported", but I will reword to make it clear.

Reviewer 2: my comment is more about how we handle updates to the field when the feature gate is on/off, not only k8s version downgrades.

Author: Hmm... the downgrade case is based on feature gate being off, so do you mean to emphasize that the field is preserved until it gets updated for the first time after downgrading?

Reviewer 2: Correct, so when the feature gate is on, then we accept setting the field, when the feature is off, we always drop the field on create, but on update there are two cases for update: 1) if it was set before, then we don't touch (we don't drop, but we also don't update it); 2) if the field was never set before, we drop it. However, since we don't allow adding to the field, then case 2 is actually no relevant to this feature.

Author: Makes sense to me.

Practically, this field can be initialized to `true` by a MutatingWebhook, and be flipped to `false`
by external integrators when certain criteria is met.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As an orchestrator developer, such as a dynamic quota manager, I have the full picutre to know when
Pods are scheduling-ready; therefore, I want an API to signal to kube-scheduler when to consider a
Pod for scheduling. The pattern for this story would be to use a mutating webhook to force creating
pods in a "not-ready to schedule" state that the custom orchestrator changes to ready at a later
time based on its own evaluations.

#### Story 2

This story is an extension of the previous one in that more than one custom orchestrator could be
deployed on the same cluster, therefore they want an API that enables establishing an agreement om
when a pod is considered ready for scheduling.

> #### Story 3

Reviewer 3: I suggest we add one more user story, relevant to @pwschuurman's suggestion (it could be 1a, but i want to make it clear the primary goal is scheduling fitness with the side effect of improving others):

```
#### Story 4

A custom workload orchestrator may wish to modify the pod prior to consideration for scheduling, without having to fork or alter the workload controller.  The orchestrator may wish to make time-varying (post-creation) decisions on pod scheduling, perhaps to preserve scheduling constraints, avoid disruption, or prevent co-existence.  

```

Author: Thanks! Will udpate.

Author: updated.

As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
to guard the schedulability of my Pods. This enables splitting custom enqueue admission logic into
several building blocks, and thus offers the most flexibility. Meanwhile, some plugin needs to visit
the in-memory state (like waiting Pods) that is only accessible via scheduler framework.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

> - **Restricted state transition:** The `schedulingPaused` state can only transit to `true` upon
> Pod's creation, and can transit to `false` afterwards once.

Reviewer 5: ```suggestion
- **Restricted state transition:** The `schedulingPaused` state can be set to `true` only when
  a Pod is created (either by the client, or mutated during admission), and can then change
  value to `false` afterwards once. You cannot change `schedulingPaused` from `false` to `true`.
```

Author: Updated.

    | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
    |------------------------|-------------------------|
    | ✅ create<br>❌ update   | ✅ create<br>✅ update    |

- **v1 guarantee:** If the feature is enabled and the new field is updated to `true`, it's still
supported to set the pod's .spec.nodeName without clearing the `schedulingPaused` flag. However,
scheduled Pods that carry `schedulingPaused=true` is a warning signal revealing one of the
following reasons:
    - The feature is not enabled in default scheduler
    - Default scheduler is at a lower version than API Server
    - A custom scheduler or administrator may have enforce setting nodeName for the Pod

> Operation-wise, you need to report this behavior to your cluster administrator.

Reviewer 5: Will any component report a metric when this situation is observed?

Reviewer 5: Is it permitted to set `schedulingPaused` to `true` _and_ set `.spec.nodeName` at create time? Would be good to make the situation clear.

Reviewer 2: > schedulingPaused to true and set .spec.nodeName at create time

we should reject that because the user intent is contradictory.

but we should allow unpausing and setting the nodeName in the same update request. With `schedulingGates`,  this means an update that results in no gates remaining can also include setting the nodeName.

Author: @sftim we had a discussion on "whether or not to allow setting nodeName for a paused Pod" back and forth. The final conclusion is no (the original version). The wording in this version is out-of-dated, sry for the confusion. I'm going to update it today.

Reviewer 5: (In terms of my feedback here, I think this is resolved).

denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->
# KEP-3521: Pod Schedule Readiness

> <!--
> This is the title of your KEP. Keep it short, simple, and descriptive. A good
> title can help communicate what the KEP is and should be considered as part of
> any review.
> -->

Reviewer 5: ```suggestion
```

ready to be considered for scheduling.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Define an API to mark Pods as scheduling-paused/ready.
- Design a new Enqueue extension point to customize Pod's queueing behavior.
- Not mark scheduling-paused Pods as `Unschedulable` by updating their `PodScheduled` condition.
- A default enqueue plugin to honor the new API semantics.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

- Enforce updating the Pod's conditions to expose more context for scheduling-paused Pods.
- Focus on in-house use-cases of the Enqueue extension point.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> We propose a new field `.spec.schedulingGates` to the Pod API. The field is defaulted to nil.

Reviewer 5: Is this true:
```suggestion
We propose a new field `.spec.schedulingGates` to the Pod API. The field is defaulted to nil
(equivalent to an empty map).
```

Author: Yes, it's true.

to guard the schedulability of my Pods. This enables splitting custom enqueue admission logic into
several building blocks, and thus offers the most flexibility. Meanwhile, some plugin needs to visit
the in-memory state (like waiting Pods) that is only accessible via scheduler framework.

#### Story 4

A custom workload orchestrator may wish to modify the pod prior to consideration for scheduling,
without having to fork or alter the workload controller. The orchestrator may wish to make 
time-varying (post-creation) decisions on pod scheduling, perhaps to preserve scheduling constraints,
avoid disruption, or prevent co-existence.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- **Restricted state transition:** The `schedulingGates` field can be initialized only when a Pod is
created (either by the client, or mutated during admission). After creation, each `schedulingGate`

> can be removed in arbitrary order, but addition of new `schedulingGate` is disallowed.

Reviewer 5: ```suggestion
can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
```

> To ensure consistency, scheduled Pod must be always carrying nil schedulingGates.

Reviewer 5: ```suggestion
To ensure consistency, a scheduled Pod must be always have empty / nil `schedulingGates`.
This means that a client cannot create a Pod that has `schedulingGates` populated in any way,
if that Pod also specifies a `nodeName`.
```

    |                                     | non-nil schedulingGates⏸️ | nil schedulingGates▶️ |
    |-------------------------------------|-------------------------|-------------------------|

>     | unscheduled Pod<br>(nil nodeName)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
>     | scheduled Pod<br>(non-nil nodeName) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |

Reviewer 5: ```suggestion
    | unscheduled Pod<br>(nil <tt>nodeName</tt>)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
    | scheduled Pod<br>(non-nil <tt>nodeName</tt>) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |
```

    > Note: if an administrator or custom controller enforce settings nodeName for a scheduling-paused
    Pod, API Server will return a 409 conflict / validation error.

- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
different enqueue plugins and wire with custom fields or conditions.

- **New column in kubectl:** To provider better UX, we're going to add a column "scheduling paused"
to the output of `kubectl get pod` to indicate whether it's scheduling-paused or not.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> - Scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller

Reviewer 5: ````suggestion
- The scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller
```

    // which schedulingGate. It's dependent on a yet-to-be-implemented fined-grained
    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
    // and needs to be consistent with how finalizes/liens work today.
    SchedulingGates []PodSchedulingGate
}

type PodSchedulingGate struct {
    // Name of the SchedulingGate.
    Name string
}
```

In the scheduler's ComponentConfig API, we'll add a new type of extension point called `Enqueue`:

```go
type Plugins struct {
    ......
    Enqueue PluginSet
}
```

### Implementation

Inside scheduler, an internal queue called "activeQ" is designed to store all ready-to-schedule Pods.
In this KEP, we're going to wire the aforementioned `Enqueue` extension to the logic deciding whether
or not to add Pods to "activeQ".

Specifically, prior to adding a Pod to "activeQ", scheduler iterates over registered Enqueue plugins.
An Enqueue plugin must implement the `EnqueuePlugin` interface to return a `Status` to tell scheduler
whether this Pod can be admitted or not:

```go
// EnqueuePlugin is an interface that must be implemented by "EnqueuePlugin" plugins.

> // These plugins are called prior to adding Pods to activeQ.

Reviewer 2: we should emphasize that those plugins need to be lightweight because they will be invoked from event handlers.

Author: updated.

can be moved back to activeQ when `.spec.schedulingGates` is all cleared
- Ensure no significant performance degradation

- `test/integration/scheduler/queue_test.go`: Will add new tests.
- `test/integration/scheduler/plugins/plugins_test.go`: Will add new tests.
- `test/integration/scheduler/enqueue/enqueue_test.go`: Will add new tests.
- `test/integration/scheduler_perf/scheduler_perf_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=BenchmarkPerfScheduling

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

Create a test with the following sequences:

- Provision a cluster with feature gate `PodSchedulingReadiness=true` (we may need to setup a testgrid
for when it's alpha)
- Create a Pod with non-nil `.spec.schedulingGates`.

> - Wait for 15 seconds to ensure it's not get scheduled.

Reviewer 5: ```suggestion
- Wait for 15 seconds to ensure (and then verify) it did not get scheduled.
```

- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature disabled by default.
- Unit and integration tests completed and passed.
- Additional tests are in Testgrid and linked in KEP

#### Beta

- Feature enabled by default.
- Permission control on individual schedulingGate is applicable (via 

> [fine-grained permissions](https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)) 

Reviewer 5: This is a private Google doc; should we hyperlink to it?

Author: hrm, I think any logged-in Google user can view it. But yes, not totally accessible anonymously.

      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

- 2022-09-16 - Initial Proposal

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> 

Reviewer 5: We should mention the `schedulingPaused` field idea here.

Author: yes, I forgot to reword this section. And also briefly explained why the alternatives are not chosen.

can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
To ensure consistency, scheduled Pod must always have empty `schedulingGates`.
This means that a client (an administrator or custom controller) cannot create a Pod that has
`schedulingGates` populated in any way, if that Pod also specified a `spec.nodeName`. In this case,
API Server will return a 409 conflict / validation error.

    |                                              | non-nil schedulingGates⏸️ | nil schedulingGates▶️ |
    |----------------------------------------------|-------------------------|-------------------------|
    | unscheduled Pod<br>(nil <tt>nodeName</tt>)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
    | scheduled Pod<br>(non-nil <tt>nodeName</tt>) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |

- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting

> different enqueue plugins and wire with custom fields or conditions.

Reviewer 2: Note that the scheduler extension points are not user facing on their own. Since the beginning we considered the framework itself and all extensions we added to it later as a "refactor", which we don't typically do behind feature gates. 

I don't mind adding a separate feature gate for this extension, but I doubt it will move the needle for the reliability of the scheduler because we can't use it to gate every side refactor we will do to implement this extension.

> different enqueue plugins and wire with custom fields or conditions.

Author: In Alpha, API field and the default enqueue plugin are disabled by default. So, it's just the backing extension point getting plumbed - which is basically a no-op for end-users in alpha, and only advanced scheduler developers would like to try it out. In other words, the extension point itself (the net portion enabled in alpha) is more an SDK that we basically introduce/refactor every couple of releases in scheduler framework.

to the output of `kubectl get pod` to indicate whether it's scheduling-paused or not.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

- The scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller
is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
in Pending state for a long time and it carries non-nil `schedulingGates`, you may find out which
component owns the gate(s) (via `.spec.managedFields`) and report the symptom to the component owner.

- Faulty controllers may forget to remove the Pod's `schedulingGates`, and hence results in
a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulabe Pods caused
by potential faulty controllers. We will evaluate necessary options in the future to mitigate

> potential abuse.

Reviewer 2: The feature itself doesn't cause creating net new pods, so I don't think we need an explicit mechanism to handle this risk. This is similar to the scheduler continuously failing to schedule pods because the cluster doesn't have enough resources (e.g., a faulty cluster autoscaler or a faulty storage provisioning controller).

I think a guiding example here is finalizers, IIRC we don't have a mechanism to deal with finalizers that don't get cleared for a while.

> potential abuse.

Reviewer 6: > I think a guiding example here is finalizers, IIRC we don't have a mechanism to deal with finalizers that don't get cleared for a while.

This is yet another thing that (in case of bugs) can result in leaking resources.
I'm not saying that we necessarily have to introduce some automatic removal of feature gates. But we need to introduce some mechanism to at least let people know that this is happening, e.g. a metric that exposes number of "notReady" pods or sth like that, so that as an operator, I can at least keep track of it.

> potential abuse.

Reviewer 2: Yeah, we have the metric from the scheduler and a default pod condition (@Huang-Wei I don't see that yet in the KEP, can we add it please).

> potential abuse.

Author: @ahg-g Yes, metric was added but I missed the default pod condition part. Updated now. PTAL.

    // and needs to be consistent with how finalizes/liens work today.
    SchedulingGates []PodSchedulingGate
}

type PodSchedulingGate struct {
    // Name of the scheduling gate.
    Name string
}
```

In the scheduler's ComponentConfig API, we'll add a new type of extension point called `Enqueue`:

```go
type Plugins struct {
    ......
    Enqueue PluginSet
}
```

### Implementation

Inside scheduler, an internal queue called "activeQ" is designed to store all ready-to-schedule Pods.
In this KEP, we're going to wire the aforementioned `Enqueue` extension to the logic deciding whether
or not to add Pods to "activeQ".

Specifically, prior to adding a Pod to "activeQ", scheduler iterates over registered Enqueue plugins.
An Enqueue plugin must implement the `EnqueuePlugin` interface to return a `Status` to tell scheduler
whether this Pod can be admitted or not:

```go
// EnqueuePlugin is an interface that must be implemented by "EnqueuePlugin" plugins.
// These plugins are called prior to adding Pods to activeQ.
// Note: an enqueue plugin is expected to be lightweight and efficient; otherwise it'd block other

> // Pods' enqueuing in event handlers.

Author: Yes, as it's a blocking call, although not in the main scheduling goroutine. Updated.

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

> ### Monitoring Requirements

Reviewer 2: @Huang-Wei perhaps we can add a label to track those pods separately in our existing metrics; we have `active`, `backoff` and `unschedulable`; we could add `notReady`

> ### Monitoring Requirements

Author: SG.

> ### Monitoring Requirements

Reviewer 3: I would prefer a label to specifically track gated for scheduling, which is not `notReady` (ready is a specific defined term, let's not overload it).  We should come up with a specific term used to refer to pods that have not closed all scheduling gates, that word should be clearly associated with the phase it is in, while ready is associated with a later phase.

> ### Monitoring Requirements

Author: How about `waiting` / `paused` / `gated`? (with an implicit "scheduling" context as prefix)

> ### Monitoring Requirements

Reviewer 3: I would be ok with `notSchedulingReady`, but we should not use `Ready` by itself in any of our APIs / metrics because that could confuse users.

"Ready to schedule" reads ok, "scheduling readiness" is a bit less clear but still ok, but in constants / fields we can't say `pod ready` / `pod not ready` without the word `schedule` or `scheduling`.  I can't think of a better word right now (`appropriate for scheduling` sounds wierd, `schedulingGated` is awkward and doesn't negate well, `allowed` is too close to rbac or policy), but there may be a better word out there if we think about it.  For right now `scheduling readiness` and fully qualified variants are ok for alpha, and we should consider better wording for this phase.

-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

- 2022-09-16 - Initial Proposal

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

- Define a boolean filed `.spec.schedulingPaused`. Its value is optionally initialized to `True` to
indicate it's not scheduling-ready, and flipped to `False` (by a controller) afterwards to trigger
this Pod's scheduling cycle.

>   This approach is not chosen is because it cannot support multiple independent controller to control

Reviewer 2: ```suggestion
  This approach is not chosen because it cannot support multiple independent controllers to control
```

specific aspect a Pod's scheduling readiness.

> - Instead of proposing a Pod field, leave the field to users (e.g., custom label/annotation)

Reviewer 2: I don't think we need to mention this as an alternative here. Building a custom scheduler is always an alternative to any scheduling feature :)

can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
To ensure consistency, scheduled Pod must always have empty `schedulingGates`.
This means that a client (an administrator or custom controller) cannot create a Pod that has
`schedulingGates` populated in any way, if that Pod also specified a `spec.nodeName`. In this case,
API Server will return a validation error.

    |                                              | non-nil schedulingGates⏸️ | nil schedulingGates▶️ |
    |----------------------------------------------|-------------------------|-------------------------|
    | unscheduled Pod<br>(nil <tt>nodeName</tt>)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
    | scheduled Pod<br>(non-nil <tt>nodeName</tt>) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |

- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
different enqueue plugins and wire with custom fields or conditions.

> - **New column in kubectl:** To provide better UX, we're going to add a column "scheduling paused"

Reviewer 3: We already have a lot of columns, I just expected this to show up in the phase column (where we summarize the state of the pod).  I don't want to specifically add another column unless we have a great reason, and this column will be empty for 99% of pods for 99.999% of their lifecycle.  Can we update the text?

will provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API

A new API field `SchedulingGates` will be added to Pod's spec:

```go
type PodSpec struct {
    // Each scheduling gate represents a particular scenario the scheduling is blocked upon.
    // Scheduling is triggered only when SchedulingGates is empty.
    // In the future, we may impose permission mechanics to restrict which controller can mutate
    // which scheduling gate. It's dependent on a yet-to-be-implemented fined-grained
    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
    // and needs to be consistent with how finalizes/liens work today.
    SchedulingGates []PodSchedulingGate
}

type PodSchedulingGate struct {
    // Name of the scheduling gate.

>     Name string

Reviewer 3: Clarify here that name must be unique among gates?  You say map earlier, but we don't clearly describe here that this must be unique among other gates.

Author: Updated.

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature disabled by default.
- Unit and integration tests completed and passed.
- API strategy test (`pkg/registry/*`) to verify disabled fields when the feature gate is on/off.

> - Additional tests are in Testgrid and linked in KEP

Reviewer 3: * Determine whether any additional state is required per gate

Author: updated.

#### Beta

- Feature enabled by default.
- Permission control on individual schedulingGate is applicable (via 
[fine-grained permissions](https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)) 
- Gather feedback from developers and out-of-tree plugins.
- Benchmark tests passed, and there is no performance degradation.

> - Update documents to reflect the changes.

Reviewer 3: * Identify whether gates can be added post-creation

Author: updated.

develop some in-house features (such as hierarchical quota) natively.

This proposal comes up with an extensible mechanics to ensure Pods are scheduled only when they are
semantically ready.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Define API to mark Pod as schedule-paused/unpaused.
- Design a new Enqueue extension point to customize Pod's queueing behavior.
- A default enqueue plugin to honor the new API semantics.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
- Update the Pod's condition if it carries `.status.schedulePaused=true`.
- Focus on in-house use-cases of the Enqueue extension point.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.

Reviewer 7: Any thoughts around an alternative way to specify the desire in the `podSpec` (potentially through a dedicated sub-resource) rather than in `podStatus`? Was thinking of something along the lines of https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#api-changes.

Reviewer 3: I was expecting this to be in spec as well.  This field is mutated prior to scheduling, and in general kubelet won’t modify spec.  It’s very similar to schedulerName and paused for me, which are both spec fields.

Would we ever parameterize or modify this value?

Did you consider schedulingPaused over schedulePaused?  Do we have any other examples of `<x><State>` naming we can look to?

Author: I chose status was b/c `pod/status` is an endpoint that would cause less conflict than `pod` itself. However, Clayton pointed out in a comment below users may set this flag in PodTemplate. In that case, `podStauts` doesn't fit anymore. I'm going to update the design.

(btw: it's a different story with preemption which is wired by `.status.nominatedNodeName` as that's always a Pod-level operation)

Regarding associating it with a sub-resource, I assume it's wired to a particular pod? like `/pods/<name>/pause`. If so, it may be of less value as we cannot update a batch of Pods' pause flag via PodTemplate. (correct me if I'm wrong).

> Would we ever parameterize or modify this value?

The value is only used when the pod is pending (.spec.nodeName=nil).

> Did you consider schedulingPaused over schedulePaused?

No strong option on the naming. `schedulingPaused` is fine with me as well.

Reviewer 7: Good point about being able to set it through PodTemplates for controllers.

Reviewer 3: Regarding parameterization, I should have been clearer:

> Will we ever have multiple possible values, or is this a true boolean?

Precedent for `paused` is true boolean, just want to be absolutely sure you don't imagine different ways of handling the pause (multiple phases, conditionals) such that this should be a) an enum or b) a struct with deeper details.

Let me think on the naming a bit more and ask for feedback from API review - I like `<verb>Paused` a bit more, but I'll want some opinions.

Author: It's a boolean, and we will prevent the flag from being flipped back and forth.

For Pods carrying `.status.schedulePause=true`, they will be "parked" in scheduler's internal
unschedulablePods pool, and only gets retried when the field is set to `false`.

Practically, this field can be initialized to `true` by MutatingWebhook, and be flipped to `false`

> by external integrators when certain criteria is met.

Reviewer 3: If a user modifies this value post creation from false to true, but before spec.nodeName is set, should we reject spec.nodeName being set?

Put another way, is this a “one way” transition or not?  If not, how will schedulers leverage / react to that transition?

Author: > If a user modifies this value post creation from false to true, but before spec.nodeName is set, should we reject spec.nodeName being set?

Do you mean when the Pod is just **_being_** scheduled, and on the same time an in-flight update (change the value from false to true) comes in? (in other cases, the pod will be updated in-place inside scheduler queues)

> Put another way, is this a “one way” transition or not?

We don't restrict the mutation. Scheduler just looks at its actual state when it's about to be scheduled. And after the Pod gets scheduled (with nodeName set), scheduler doesn't care about that Pod anymore.

Author: > I think this should be a one-way transition from true to false. The field can be set to true only on create and never on update.

This would make our life easier - we don't need to worry about a being-scheduled Pod gets updated elsewhere with the paused flag from false to true.

My concern is: by restricting the transition, do we lose some flexibility? like, is the "reviving a pod" case mentioned in https://github.com/kubernetes/enhancements/pull/3522#discussion_r973441963 is a valid usecase? Want to hear more other real users. cc/ @wilfred-s

Reviewer 8: If pod is waiting for scheduling, we can mutate the `spec.schedulePaused`, if pod is in scheduling, we disallow the transition.

Reviewer 2: > If pod is waiting for scheduling, we can mutate the spec.schedulePaused, if pod is in scheduling, we disallow the transition.

It will be hard to be this granular because the scheduler and api-server don't communicate via shared memory, the state is communicated via the object itself and so it will be hard to avoid race conditions. I would start strict and relax later if the need arises.

Author: > if pod is in scheduling, we disallow the transition.

This needs to yield a signal to API Server from scheduler, and the only way to achieve this is through some API object's change. But I don't think it's practical to come up with another level of abstraction on the scheduling phase for a Pod.

Author: @ahg-g for unschedulable Pod, it makes sense to restrict the transition from true->false. But for scheduled Pod, let's brainstorm in https://github.com/kubernetes/enhancements/pull/3522#discussion_r974485890.

Author: Fold this thread as we almost reach a consensus in the thread of https://github.com/kubernetes/enhancements/pull/3522#discussion_r974485890.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As an application developer, I'm knowledgeable to know when my Pods are schedule-ready, but I know

> nothing about scheduling internals. It'd be ideal for me to only set/unset the "paused" flag,

Reviewer 3: Can I set this value on any pod?  Any workload object with pod templates?  Why or why not?

Reviewer 2: I think in most cases this field will be set to `true` in a pod-level webhook. But I can't think of a reason why we would want to prevent setting it on workload objects.

Author: On any pod, yes. PodTemplate, yes (will update the design to add the field in PodSpec so it can be changed in PodTemplate).

instead of developing scheduler plugins, to control the schedulability of my Pods.

#### Story 2

As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins

> to guard the schedulability of my Pods.

Reviewer 2: I think we need to clarify why the pauseScheduling flag is not enough to address this story. One could argue that an external controller could be built to set the pauseScheduling to "guard the schedulability of my pods"

Author: Updated PTAL.

Reviewer 9: For the motivation #2 and #3 in the discussion doc (https://docs.google.com/document/d/1cPxF3wIbGRw_inv4CCG6z6qC4WjHAPR5GLT__dRBx7I/edit#heading=h.whhghc8mxakf), how would a user guard schedulability of pods on independent conditions? An API that allows for some dynamic conditions (like an array of scheduling guards, similar to pod readiness gates design) would allow for this, rather than a single boolean.

> One could argue that an external controller could be built to set the pauseScheduling to "guard the schedulability of my pods"
I agree, a user can create a unified controller to control flipping the flag (eg: ensure that all dependent secrets are available, and dependent PVC is bound), but that makes the logic dependent, and prevents the logic from being distributed. Say the user wants to pause scheduling on two conditions, 1) PVCs are bound and 2) a custom condition related to networking configuration being available to the cluster. The first condition is fairly generic and can apply to any Stateful workload, the second is custom user logic.

Author: @pwschuurman are you supportive of the 2nd story or not? I guess yes :)

Just as you said, some advanced developers may choose to ignore the binary flag, and implement their own combinations of enqueue plugins, which is exactly the motivation of the 2nd story. In other words, in this story, you don't need the API field to achieve the goal. But this needs extra expertise on scheduler framework, so the first story and API field are still necessary.

Generally, story 1 and story 2 are complementary to each other to fit different personas and use cases.

Reviewer 9: I see, so for the 2nd story, an advanced user could leverage the Scheduling Plugin API today (as long as their Kubernetes cluster supports the Plugin Scheduling Framework).

> Generally, story 1 and story 2 are complementary to each other to fit different personas and use cases.

And just to confirm, a user following the 1st story using this API field can complement existing Scheduling plugins set up by say a Kubernetes cluster administrator (eg: an existing plugin installed on the cluster as a default).

Author: That's correct. The general rules to how to cooperate with different enqueue plugins are documented in the Design#Implementations section, some teeny tiny details may need to be polished/reviewed in the future code PR though.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

<<[UNRESOLVED]>>

The feature will be guarded by a feature gate and disabled by default in alpha. When alpha, surely 
the Pod API field needs to be disabled, but I'm open to hear feedback on whether to disable scheduler's
ComponentConfig API as well as the extension point logic.

If we disable it, then users cannot use the new extension point at all.
If we don't disable it, people can still customize their enqueue behaviors, like wire it with a
customized field (like an annotation key).

> 

Reviewer 2: The feature gate should probably be only on the Pod api not the new scheduler extension point. I don't think we ever used feature gates while developing and evolving the scheduler framework since it was always thought of as a refactor that has no user visible effects.

Author: Yes, that's exactly what I thought of. Will let it soak for a while, and then remove UNRESOLVED to clarify the semantics for extension point.

<<[/UNRESOLVED]>>

<<[UNRESOLVED]>>

As we don't write unschedulable condition to the Pod's spec if it carries `.status.schedulePaused=true`,

> it may provider better UX to add a "paused" column to the output of `kubectl get pod (-o wide)`.

Reviewer 2: +1

Author: let it soak for a while, and then I will clarify by removing UNRESOLVED.

<<[/UNRESOLVED]>>

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
should provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API

A new API field `schedulePaused` will be added to Pod's status:

```go
type PodStatus struct {

>     SchedulePaused *bool // <-- New field.

Reviewer 7: Will the field become non-mutable for a pod that has finished scheduling and has a nodeName assigned?

Reviewer 3: In the context of a pod which can only be scheduled once, i would probably argue the expected behavior is the field is immutable after nodeName is set (prevent unexpected transitions early).

I had a question above about the behavior prior to scheduling that as similar

Reviewer 2: I think validation should be even tighter on updates and only allow true -> false transitions, otherwise the risk of race conditions is much higher.

Author: Semantically, if a pod's `.spec.nodeName` is non-nil, it means the Pod has passed scheduling phase, so `schedulingPaused` is a no-op. So what's the net benefits of making it immutable after the pod's nodeName is set?

Given scheduled Pod can still get rejected by kubelet, esp. when multiple scheduler are in place, so in API Server's perspective, if we make the paused flag immutable, can we unset its nodeName AND flip the paused flag in one API call? (my guess is no)

Author: > Restricting the transition of the flag from true-to-false will help with that.

> In fact, another thing we should do is to reject nodeName update if the flag was set to true.

The case I described above was primarily the conflict among multi schedulers, so no matter w/ or w/o the true->false restriction, a scheduled Pod can somehow be rejected by kubelet. My point was, ops was able to "reprieve" this Pod by unsettling its nodeName (to nil) and flip schedulingPaused (to true), but with this restriction they cannot - deleting and recreating the Pod is the only way.

We may argue this is a hypothetical case, but better discuss here. If we decide not to support this, we should document it well.

Reviewer 2: > The case I described above was primarily the conflict among multi schedulers, so no matter w/ or w/o the true->false restriction, a scheduled Pod can somehow be rejected by kubelet

I think we shouldn't use that case as a reference and a design pattern to follow. Multi schedulers is supported, but the recommended pattern should be for each scheduler to handle different workloads on different nodes (we need to make that clear in the docs if we don't already). So even for multi schedulers, Kubelet admission logic is acting as a last resort for protection, and ideally shouldn't be rejecting pods if the multiple schedulers where setup correctly.

> My point was, ops was able to "reprieve" this Pod by unsettling its nodeName (to nil) and flip schedulingPaused (to true), but with this restriction they cannot - deleting and recreating the Pod is the only way.

I would argue that nodeName should also be set once, but I don't think we can tighten validation now on this since it would be backward incompatible.

Author: A consensus we may reach is the one-time transition of _unscheduled_ Pod - you cannot flip the .schedulingPaused field via update back and forth, only from true->fase:

|                                     | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
|-------------------------------------|------------------------|------------------------|
| unscheduled Pod<br>(nil nodeName)   | ✅ create<br>❌ update   | ✅ create<br>✅ update   |

In terms of _scheduled_ Pod, I rethink a bit about the case mentioned above, to "revive" this Pod, we could "achieve" via:
- 1 API call: update **both**  nodeName (to nil) and schedulingPaused (to true), but this path is assumed to be blocked (❌ above) already.
- 2 API calls: firstly update schedulingPaused (to false), and then update nodeName (to nil). In this path, the first step is pretty misleading, as you'd reach a state that a scheduled Pod with schedulingPaused=true. So although the second call is allowed so far, the first API call looks problematic, and overall, we may abandon this path.

With the above being said, I think we should impose validation of two state changes:

- disallow a Pod's update on changing schedulingPaused to true.
- disallow any state that a Pod becomes schedulable while also carrying schedulingPaused=true.

Validation diagram is like this:

|                                     | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
|-------------------------------------|------------------------|-------------------------|
| unscheduled Pod<br>(nil nodeName)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
| scheduled Pod<br>(non-nil nodeName) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |

Author: I will update the design if ^^ makes sense to all of you.

Reviewer 3: Re: blocking setting nodeName when scheduling is paused, I'm a bit concerned about blocking that because today a client is allowed to forcibly schedule a node, and that may not be a scheduler, but might be an admin.  Also, schedulers that are not aware of the scheduling paused field (have not been code updated) should behave predictably (i.e. we have to decide what the behavior is) when this field is added and a user uses it.

Scenario:

1. Cluster has two schedulers, default and custom
2. Some automation uses the custom scheduler to create pods
3. Cluster is upgraded to support `schedulingPaused`
4. Default scheduler updated to understand schedulingPaused, custom is not
5. A new workload of setting paused is added to the default scheduler (i.e. new consumer)

Can we make arguments for or against:

1. Allowing a human admin to force schedule a paused pod via setting nodeName without being aware of paused?
2. Preventing the custom scheduler from being broken by a human setting "paused" on one of the custom workload jobs (which puts the scheduler into a break loop)?

There is also the forward compatibility angle of adding this field for the "unaware" custom scheduler - we need to make an argument for or against support there from "does the addition of this field have unexpected outcomes for existing code that is unreasonable / dangerous"?

Author: > 1. Allowing a human admin to force schedule a paused pod via setting nodeName without being aware of paused?
> 2. Preventing the custom scheduler from being broken by a human setting "paused" on one of the custom workload jobs (which puts the scheduler into a break loop)?

I think we can have an in-tree admission controller to unset the "paused" field for scheduled Pod. cc/ @ahg-g 

The intention of "blocking setting nodeName when scheduling is paused" was to avoid discrepancy, which is "a scheduled pod, but carries a schedulingPaused state, which looks weird". If we can clear the "paused" field, we're good to always guarantee a consistent view.

Reviewer 3: >  "a scheduled pod, but carries a schedulingPaused state, which looks weird"

Weird is vague :) - is this a better way to describe what you mean?

> Will clients be confused if it remains set?

How would the scheduler have to react in that case?  Should all scheduler impls have to track both fields?  Is a scheduler allowed to ignore "paused" (perhaps one that has not been upgraded yet)?  Forward compat was the reason that I was originally asking.

Will server side apply or other client apply loops be confused if the field remains set?  I.e. "paused" is the initial value, will a subsequent apply work as the user would expect?

Will a workload controller (statefulset, deployment) pod template transitioning from `schedulingPaused: true` to `schedulingPaused: false` result in redeployment of the pods underneath it?  If it doesn't, will the clearing of paused by the scheduler impact how the various workload controllers deploy pods?

> we're good to always guarantee a consistent view.

Consistency is good, but it's generally a default and can be overriden by ergonomics, forward compatibility, and careful consideration of the impact of adding the field.  We don't have a great set of rules for "how to think about adding fields to core objects like pods safely" and I'm trying to see if we can tease those out here in that absence.

Author: > Weird is vague :) - is this a better way to describe what you mean?

Semantically, a pod that has nodeName assigned indicates it has passed scheduling, so it doesn't make sense to have a pod in that state. 

We're going to add a column "scheduling paused" to `k get pod`. So if in the output of `k get pod`, it would show a pod has node name assigned, and also schedulingPaused=true, that would confuse the user a lot IMO.

> How would the scheduler have to react in that case? Should all scheduler impls have to track both fields? Is a scheduler allowed to ignore "paused" (perhaps one that has not been upgraded yet)? Forward compat was the reason that I was originally asking.

Default scheduler itself doesn't track the "paused" state if it's already scheduled. However, this doesn't mean we _should_ leave it this way. Integrators may rely on this field to compose their behavior. For example, when we built preemption feature, `.status.nomiantedNodeName` is set to indicate a pod is preempting other pods, and that field is supposed to be cleared when it gets scheduled. Well, in scheduler's POV, we _could_ claim we don't care about that state if it's scheduled (by default scheduler or by custom scheduler or enforced by admin), but Cluster Autoscaler does rely on that field to proceed their scale-up/down logic. This is the same case for "paused" state - make the state always accurate either by (1) disallowing update nodeName for "paused" pod or (2) actively clear the "paused" state for scheduled pod.

> Will server side apply or other client apply loops be confused if the field remains set?

Given we're gonna enforce a true->false on update, user would see the update as expected.

> Will a workload controller (statefulset, deployment) pod template transitioning from schedulingPaused: true to schedulingPaused: false result in redeployment of the pods underneath it? If it doesn't, will the clearing of paused by the scheduler impact how the various workload controllers deploy pods?

No, it doesn't yield redeployment. And it won't impact how the various workload controllers deploy pods either.

(BTW: scheduler doesn't clear the state, only controllers do. Or, "to-be-determined" admission controller in a particular circumstance.)

Reviewer 2: > I think we can have an in-tree admission controller to unset the "paused" field for scheduled Pod. cc/ @ahg-g

I don't think we want to go that route, this field is not a status field to reconcile it in this fashion. 

What I was hoping to achieve with "blocking setting nodeName when scheduling is paused" is consistency on scheduling **behavior** across all schedulers and scenarios, meaning a pod will in all circumstances will not be assigned a node if it was paused. 

@smarterclayton consider the scenario where a feature is built using this flag, like statefulset migration, and we have a customer scheduler that wasn't updated to take this flag into account; then we have to select between two failure scenarios:

1) the statefulset migration feature being broken in an unexpected way if we don't block setting nodeName when pause=true
2) custom schedulers erroring out

I think the second failure is more tolerable because it doesn't put the cluster in an unexpected state.

Author: > I don't think we want to go that route

That was an idea based on supporting the scenario Clayton mentioned - a custom scheduler or admin force setting a nodeName while it's paused.

> I think the second failure is more tolerable because it doesn't put the cluster in an unexpected state.

+1.

Reviewer 2: but I wouldn't push hard on it if others consider this backward incompatible.

}
```

In the scheduler's ComponentConfig API, we'll add a new type of extension point called `Enqueue`:

```go
type Plugins struct {
    PreFilter PluginSet
    ......
    Enqueue PluginSet // <-- New field.
}
```

### Implementation

Inside scheduler, an internal queue called "activeQ" is designed to store all ready-to-schedule Pods.
In this KEP, we're going to wire the aforementioned `Enqueue` extension to the logic deciding whether
or not to add Pods to "activeQ".

Specifically, prior to adding a Pod to "activeQ", scheduler iterates over registered Enqueue plugins.

> An Enqueue plugin must implements the `EnqueuePlugin` interface to return a `Status` to tell scheduler

Reviewer 2: ```suggestion
An Enqueue plugin must implement the `EnqueuePlugin` interface to return a `Status` to tell scheduler
```

whether this Pod can be admitted or not:

```go
// EnqueuePlugin is an interface that must be implemented by "EnqueuePlugin" plugins.
// These plugins are called prior to adding Pods to activeQ.
type EnqueuePlugin interface {

> 	  Plugin

Reviewer 2: indentation

    PreEnqueue(ctx context.Context, state *CycleState, p *v1.Pod) *Status
}
```

> A Pod can be moved to activeQ only when all Enqueue plugins returns `Success`. Otherwise, it's moved

Reviewer 8: Move the pods rejected by default enqueue plugin back to backoffQ/activeQ seems useless when the `spec.SchedulePaused` remains in true. We should add some filter logic at that moment.

Author: You mean the current logic of `flushUnschedulablePodsLeftover`? Yup, we should tweak the impl. logic a bit. But in terms of KEPing, it's way low-level impl., and also someday we may remove it, I suppose we shouldn't specify the details in this doc.

develop some in-house features (such as hierarchical quota) natively.

This proposal describes APIs and mechanics to allow users/controllers to control when a pod is
ready to be considered for scheduling.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Define API to mark Pod as schedule-paused/unpaused.
- Design a new Enqueue extension point to customize Pod's queueing behavior.
- A default enqueue plugin to honor the new API semantics.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
- Update the Pod's condition if it carries `.spec.schedulePaused=true`.
- Focus on in-house use-cases of the Enqueue extension point.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> We propose a new field `.spec.schedulePaused` to the Pod API. The field is defaulted to `false`.

Reviewer 10: I was thinking (and I suggested it to @ahg-g offline just now) that it might be worth making this a list of strings, in a similar fashion to finalizers.

Once the list is empty, the pod is allowed to be scheduled.

This allows multiple controllers/webhooks to decide whether a pod can be scheduled based on different criteria.

It might be useful to have an endpoint just for this field (and disallow edition via general patch/update) so that there are less chances that a controller accidentally removes all the strings. Or maybe we can just  have a simple update validation: only one string can be removed at a time.

Although, for finalizers we don't do that. The only related thing is an RBAC rule that allows controllers to just modify finalizers but nothing else.

Reviewer 2: @pwschuurman this should help simplify https://github.com/kubernetes/enhancements/pull/3336

Author: Thanks @alculquicondor for this idea. This looks more extensible, and we may name it
`ScheduleReadinessGates []string`.

I'm interested to learn from API experts if there are any traps in adopting this finalizers-like API pattern.

Reviewer 8: It's great, then each controller can have its own enqueue plugin if it needs.

Reviewer 10: Actually, the whole point of the list is that developers don't have to implement scheduler plugins. They just control when to remove their own string from the list.

Reviewer 4: random thoughts from finalizers:
* is ordering important or can any gate in the list be removed at any time?
* is it important to protect against one gate owner messing with a gate belonging to another other (either accidentally or intentionally)? finalizers don't enforce any additional permissions by default (the OwnerReferencesPermissionEnforcement admission plugin that enforces this is optional and off by default), so any API client with update permission can add/remove any finalizer
* what is the expectation for gate owners to report status when blocked? is there any concept of a timeout?

Author: Re "ordering": ordering is not important. Gates can be removed after the pod gets created, uni-directionally. (i.e., you cannot append/remove back and forth).
Re "permission": each gate owner is _supposed_ to be a good citizen and only remove their gate(s) when needed. Permission protection is a good-to-have feature, but I don't see it as a blocker at least for alpha.
Re "timeout": I guess it's the same story with finalizers. If some gate owners get disconnected, then the gates cannot be cleared and thus stay in that state. There is no timeout concept on the scheduler side, so totally relies on gate owners to clear the gates.

Reviewer 4: > Permission protection is a good-to-have feature, but I don't see it as a blocker at least for alpha.

If you don't have it initially, I'm *really* skeptical it will ever get added. That's what happened with finalizers, and it's why there's no compatible way to walk back that feature's wide-open posture

Author: Gotcha. We have to make it right in the very beginning if we go with this approach.

Reviewer 2: +1 to gate it from the beginning if there are the mechanics to do it.

Reviewer 2: I assume we will need to setup a endpoint and the path would include the string to be removed/added so that we can setup RBAC on a per string basis. Is this feasible @liggitt ?

Author: @ahg-g I'm not sure RBAC supports a so-called "dynamic" endpoint/path.

IIUC, we may need to come up with something like `ScheduleReadinessGates []OwnerRefs`, each ownerRef needs to be GVRed, so that RBAC can tell if the operation is allowed.

Reviewer 2: Don't we support something similar for namespace access?

Reviewer 2: It turns out that a couple of other KEPs are also stuck at this same decision point and there is a discussion going on regarding the path forward [1]. To unblock this KEP, Jordan is suggesting to "make a generic reference to subdividing permission/access with a yet-to-be-determined mechanism consistent with liens and finalizers". 

[1] https://groups.google.com/g/kubernetes-sig-auth/c/U7rQiOEsnmA/m/aU82M1aKCgAJ?utm_medium=email&utm_source=footer

Reviewer 4: at first glance, these seem similar to custom pod readiness gates, where a list of conditions can be registered in the pod spec, and until all those conditions are present/true, the pod is not considered ready

would a parallel list of "scheduler gates" make sense here? they would be registered in `spec.schedulerGates`, the external actors would update pod status to add their conditions, and once they were all present/true, the scheduler would fire

that has the benefit of:
1. matching custom readiness gates approach
2. giving an existing status field / mechanism for actors to publish their individual condition
3. giving an existing status field for those actors to communicate updates / negative status (register their condition with a value of False and a reason, etc)

Reviewer 11: whether it is a list of readiness gates or a boolean field I am concerned with defaulting to false which changes the current behavior where pods are considered ready to schedule by default.

Also mutation rules need to be spec-ed. For the same of the next part let us assume it is a a boolean field. Here are the rules:

1. `false` => `true`: OK, Pod was waiting on something that was availed.
2. `true` => `false`: NOT OK, Pod was ready and thus may have been already somewhere in the scheduling, starting process and can not be brought be back `waiting for something that was availed`

Author: @khenidak you're understanding right actually. It's just the field is named "schedulingPaused", so defaulted to false, indicating it's scheduling ready by default. The rules you mentioned are also correct, just swap `false` and `true`.

Reviewer 11: one more thing `paused` typically means ready for `play` and `pause` again. Should consider `.spec.schedulerReady` or something that does not implicitly imply that the state machine can move backward and forward.

Author: @liggitt if we go with `spec.schedulerGates`, I suppose the following problems are hard (or impossible) to tackle?

1. we cannot gate which user can control which schedulerGate as any condition can be updated as long as you have update/patch permission on pod/status?
2. we cannot easily enforce a uni-directional transition for a particular ready condition?

Reviewer 3: I think another way to ask Jordan’s question is about the future:

“Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future”

If we can’t say yes, the boolean is appropriate.  If we *could* say yes, we should enumerate what our arguments for or against those briefly here.

Reviewer 3: I’m a little hesitant to endorse a workflow conceptual field without very strong and immediate use cases that justify it.  We have always tried to find the minimal declarative structure with maximum composability - even simple lists of strings are rarely sufficient for end-user (vs system) control.  So the bar would be very high to add such a field now.

Author: > “Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future”

I don't think so. Each gate is supposed to represent a particular condition, and behave statelessly.

Author: > So the bar would be very high to add such a field now.

@smarterclayton to clarify, by "such a field" you meant the `[]string` field brainstormed in this thread; instead of the originally proposed `bool` field, right?

Reviewer 10: I'm in favor of @liggitt's idea

Reviewer 10: > Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future

Actually, yes.

@pwschuurman mentioned a use case related to migrations. The original motivation for this KEP was about queueing. What if users wanted to have both behaviors? The different controllers gating a pod to be scheduled need to coordinate.

Reviewer 3: Thinking about this last night, one of the advantages of a readiness gate approach is that it allows a human to create pods that won't be scheduled yet (through either an unknown gate or an explicitly predefined gate).  So that's two use cases worth calling out (including the stateful set orchestrator):

1. A workload orchestrator can use a scheduler readiness gate to perform custom control of when pods are released to scheduling (statefulset use case)
2. A human can create a pod and delay scheduling until such a time as other resources are initialized iteratively, or allow it

Both of those bring up key design considerations:

1. How should a workload controller handle pods that remain in scheduling unready due to paused for a long-time?  Do we need to do anything additional over "unable to schedule"?  Should the scheduler have a time based window before a paused pod is flagged?
2. Can a cluster be brought down by large numbers of unscheduled pods?  We have a component responsible for lifecycling completed pods (pod GC), but not one for unscheduled pods.  Are there additions we need to make to quota or alerting to help administrators manage this list?  (apologies if this is in KEP and I missed it)

Author: Very good points on the considerations.

Re 1: for now, I don't quite consider we should impose a time based mechanics to dishonor the API (paused state), even conditionally. The paused state is supposed to be a strong scheduling gate until it's cleared.
Re 2: it does make sense, esp. in operation's view, to prevent this feature from being misused - maybe due to faulty controllers. But I'm hesitant to add support from the beginning for the hyperthetical requirement. I'd like to document this potential behavior for now, and revisit once we get some real feedback. Maybe associated with a PDB-like mechanics.

I will update the KEP to describe the potential symptoms.

Reviewer 2: Regarding 1: I also don't think we should impose time limits for two reasons:
1) currently we don't have time limits for pods that fail to schedule
2) I don't think this feature will encourage creating more pods and so it shouldn't really impact scalability in a negative way; in fact, this feature enables a new design pattern that should improve scalability: currently the scheduler continuously attempts to find a node for unschedulable pods, and so wasting scheduling cycles; this feature gives orchestrators a mechanism to tell the scheduler to ignore those pods until a later time, and so practically saving scheduling cycles.

Regarding 2, we can add a quota scope to limit the number of "not-ready to schedule pods": https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-scopes

Reviewer 3: > currently we don't have time limits for pods that fail to schedule

I should have been more clear with "time limit", I was asking two questions together (the first you answered), the second being

How long before we set the status condition on a paused pod with a reason for why the pod is not scheduled?  An old client may not even be able to observe the paused field (for instance, a web interface like octant or a cli that doesn't use server side get), and so to those users a paused pod is indistinguishable from "scheduler has hung / is unavailable".  

Re: 2, I think you are correct that this feature is net neutral on number of pods (actually, slightly reduces the aggregate number of pods) because it create no more pods than would normally get scheduled.  If someone attempted to implement reservation via creating a pause pod they will have to update or delete it and recreate it to use the reservation.

Author: > How long before we set the status condition on a paused pod with a reason

Scheduling-wise, we can separate the conditions into two categories:

- [PodScheduled](https://github.com/kubernetes/kubernetes/blob/8f269d6df2a57544b73d5ca35e04451373ef334c/staging/src/k8s.io/api/core/v1/types.go#L2656) - for paused pods, as mentioned in Non-Goals section we won't update this condition, as it stands for a "scheduler tried but failed", and it's used by CA heavily to determine whether the pod needs a new node to accormordate.
- to-be-determined condition type like "PodSchedulingPaused" - I'm not quite sure if we need it as it seems this is for lower-version clients only. My understanding is even with this new condition, the client (UI/cli) still cannot understand it without updating their own logic, then given code changes are needed anyway, is there a big difference with upgrading to a new client-go version so the paused field is aware?

Reviewer 2: Right, the exiting `PodScheduled` condition carries the meaning that the scheduler attempted to find a node and failed, which currently triggers autoscaling in cluster-autoscaler. There are two things we can do here:

1) CA should start to recognize the flag
2) We could use the same condition, but use a different `reason` since `reason` is supposed to be machine readable. This will also require a change in CA, but it is a change that we can patch back to older CA versions (for example, the change would be to ignore the pod if the `PodScheduled` condition is false and reason is "SchedulingNotReady")

Author: > we can patch back to older CA versions

Maybe we just make changes to the latest CA version as CA [suggests](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#releases): "We don't do cross version testing or compatibility testing in other environments."

Reviewer 3: If we're worried about CA and others reacting to `PodScheduled=false`, we *could* set `PodScheduled` to `Unknown` after some time with an appropriate reason, and then the server side printer for pods could leverage that if we had to.  I think the hard part would be trying to decide how long to wait (15s? 30s?), and again, whether it is worth it.

It may be simpler to update the server side printer for pods to output "waiting" when scheduling is paused (in which case the condition is less critical).

Author: The original plan was to not set the status; instead, exposing the Pod's state by events, metrics and CLI support.

That was based on not breaking lower versions of CA. However, if CA is guaranteed to only support the exact matched k8s version, update a pod's status as `PodScheduled=paused` doesn't sound like a bad idea. In that case, we don't need to rely on a timeout - internally the paused pod will still go through a lightweight scheduling process (enqueue extension point) where we can update the status.

Reviewer 3: We could also have creation default the pod Scheduled condition to false and reason `WaitingForGate`.

Reviewer 2: > We could also have creation default the pod Scheduled condition to false and reason `WaitingForGate`.

Make sense to me, this saves us a mutating request compared to having the scheduler doing it. But I am not sure if we should use condition status `unknown` or `false`. I prefer `false` if we can guarantee compatibility with CA,  @x13n what do you think?

Author: > We could also have creation default the pod Scheduled condition to false and reason WaitingForGate.

That's fair. Just to clarify: the condition "PodScheduled=WaitingForGate" is applied when it goes through the lightweight scheduling process, instead of defaulted by `pkg/apis/core/<something>`.

Reviewer 2: No, we can default it at pod creation.

Author: seems we don't have precedence of defaulting a condition? (I checked pkg/registry/core/pod/strategy.go)

But, if API folks are fine with the pattern that at least save the scheduler's efforts.

Reviewer 3: We're adding a Kube feature, defaulting can do many things, while it's novel, it's certainly worth trying.  If we come up with a reason to remove it before beta, that's low cost, but we do have defaulting of status (namespace status phase defaulting).  I think it's a nice usability win, and we can record "will look for implications of novel condition defaulting for GA".  We may only be able to do the defaulting during creation (can't do it during update), so that's a possible wrinkle.  Could simply list it as a run-at for mitigation.

Author: > I think it's a nice usability win

+1. Already added it in the monitoring section.

several building blocks, and thus offers the most flexibility. Meanwhile, some plugin needs to visit
the in-memory state (like waiting Pods) that is only accessible via scheduler framework.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
or not. This enable scheduler plugin developers to tryout the feature even in Alpha, by crafting
different enqueue plugins and wire with custom fields or conditions.

- **New column in kubectl:** As we don't write unschedulable condition to a Pod's status if it carries
`.spec.schedulePaused=true`, to provider better UX, we're going to add a column "schedule paused"
to the output of `kubectl get pod` to indicate whether it's scheduling paused or not.

> - **Restricted state transition:** The `schedulePaused` state can only transit to `true` upon

Reviewer 11: Thank you!. my earlier comment brought this up.

  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

- Upgrade
  - While the feature gate is enabled, `.spec.schedulingPaused` can be mutated by users.
  - While the feature gate is enabled and `.spec.schedulingPaused` is not set, a default value will
  be given to behave the same like it's specified by users.
- Downgrade
  - Previously configured values will be ignored.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> N/A. The feature doesn't affect nodes.

Reviewer 11: True. But a situation where api-server is vCurrent+1 and scheduler is vCurrent things will be unpredictable specially for inexperienced users.

As we stand today i don't think we have a solution for this. Normally it has no impact on the behavior because old clients don't see new fields but this field is of high significance (high expectations when set to `true`). 

I think we will have to clarify the documentation specifically covering scheduler v api-server version skew for upgrade and downgrade.

Reviewer 3: I would go even further:

> New fields must be additive.  The new paused field, due to forward compatibility requirements, may be ignored by custom schedulers for long periods.  Our default scheduler will ignore it unless (configuration / gates / version).

"paused" isn't special, schedulers may ignore it, and we can't break custom schedulers through its introduction.

We *should* do what we can to make administrators aware that the feature is not implemented or enabled - that might be something like a metric on "pods scheduled without clearing schedulingPaused explicitly", as well as warnings (possibly) reported to clients that do the scheduling without clearing that value.  But it will always be supported to set `spec.nodeName` without clearing `schedulingPaused`, and it should be possible as you note to discover whether a component is respecting paused.

Author: > we can't break custom schedulers through its introduction

@ahg-g I think this is the hard requirement of v1 guarantee. In other words, if API Server is at vCurrent and the feature is enabled, and scheduler at vCurrent-1, and it happens that some controller sets the paused flag, we need to ensure this pod get scheduled properly. (The scheduler can be custom scheduler or admin). So, like Clayton mentioned, it's a must to:

> But it will always be supported to set spec.nodeName without clearing schedulingPaused

So our discussion on erroring out "setting nodeName on a paused pod" is not an option at all.

Author: > We should do what we can to make administrators aware that the feature is not implemented or enabled

Metric works in this aspect. Default scheduler can generate some logs when it detects some scheduled pod carries the paused field. Other than that, it seems there is no good way to "warn" the admin/user. Maybe we can dump a warning in kubectl's output to alert on scheduled Pod carrying paused state?

Reviewer 3: Reading this again this morning, I want to nuance my statement and define what constitutes "break". I think @ahg-g makes good points in the other thread, and I'm trying to find the right use case to guide what constitutes a breakage (I may be thinking too strictly):

TL:DR: I think blocking `spec.nodeName` when `schedulingPaused` is safe because the use of schedulingPaused is opt-in for users - and we can't set schedulingPaused automatically until the feature is in use for 1 release.

If a user (human or workload controller) creates a pod without setting scheduling paused (regardless of whether this feature is on), they expect the pod to be scheduled as long as it fits on a node.  An administrator may override the default user-observable behavior of scheduling by: adding admission hooks that assign a non-default scheduler, adding admission hooks that alter the object to make it not fit, or customizing the default scheduler to have different behaviors.  These are all administrator defined variants that are allowed deviations.  A user expects scheduler to fit or not quickly, but there is no hard time boundary to schedule.  Users cannot observe the inner workings of the scheduler except as the scheduler chooses via status conditions.  Ultimately the only breakage a user will notice is a pod that sits forever in schedulable without some reason being given via the Schedulable condition - so we have wide latitude to change the scheduler flow by default.

We expose to a scheduler (default or custom) a similar contract - they observe a pod via the API in the `Schedulable` lifecycle state and are expected to decide which scheduler has responsibility based on a) default configuration and schedulerName, and b) administrative configuration.  Schedulers may make arbitrary decisions on scheduling, including using spec / metadata fields (or schedulerName) to push pods through workflows (i.e. schedulerName -> `custom/measure` -> `custom/wait_for_reservation`).  Such a scheduler must not be broken by the addition of a new lifecycle state, and in general custom schedulers should not break *if the feature is not leveraged*.

The addition of fields that alter the behavior of the scheduler for pods in the `Schedulable` lifecycle state is not new - for instance topology constraints, tolerations, and custom resources all required the user or admin to opt-in to leveraging those features through explicit action (by setting the field, configuring taints on nodes, or adding admission hooks that defaulted values).  Historically we have said the scheduler simply ignores those fields until updated, and then begins enforcing them (e.g. during an upgrade from alpha to beta, where the gate becomes available).  This is generally understood to be "not common enough for us to care about" :)

The behavior about blocking `spec.nodeName` being set if `schedulingPaused` is effectively `true` is an added capability above this - the user *has* opted in (and `schedulingPaused` is either user set or admin set, just like any other constraint), but we can enforce the logic at the apiserver level.  The scheduler that isn't aware of `schedulingPaused` and is now prevented from scheduling is only blocked due to the user's action (setting schedulingPaused) vs the behavior of the upgrade.

What we cannot do is set schedulingPaused by default on all new pods during a Kube upgrade without admin or user intervention.  That *would* break those unknowing schedulers.  An admin is allowed to enforce that (perhaps via an opt-in config flag), but we cannot by default without changing the observable behavior of the pod `spec.nodeName` field.

Author: > The scheduler that isn't aware of schedulingPaused and is now prevented from scheduling is only blocked due to the user's action (setting schedulingPaused) vs the behavior of the upgrade.

I _thought_ this is part of v1 guarantee..

> What we cannot do is set schedulingPaused by default on all new pods during a Kube upgrade without admin or user intervention.

No, we won't.

> we can't set schedulingPaused automatically until the feature is in use for 1 release

Even the feature reaches beta/GA, we won't make pods paused by default - it will still be opt-in feature.

---

> Ultimately the only breakage a user will notice is a pod that sits forever in schedulable without some reason being given via the Schedulable condition

This may be something worth discussing (also mentioned a bit in https://github.com/kubernetes/enhancements/pull/3522#discussion_r978872725), _if_ we expose a new reason on an existing condition type (`PodScheduled`), as you mentioned, the clients would have context to understand what's going on here, I read it as a good-to-have; however, if we consider some integrators like CA, from their perspective, they just look at the condition `PodScheduled` to treat it as unschedulable, and try to spin up a new instance. So with this regard, this will break their workflow.

Reviewer 2: Indeed, this is an opt-in feature and will never be defaulted to "not ready" unless through a webhook, which is admin controlled. This feature shouldn't really be used in a cluster with a custom scheduler that doesn't support it yet, and I think it is appropriate for the api-server to be strict on that by rejecting setting nodeName when the pod is not ready.

As you mentioned, ideally all scheduling constraints should be this way, but unfortunately we can't enforce them via the api-server without introducing significant overhead.

Reviewer 3: Sounds like we're all in agreement then, we should fail the attempt to set `spec.nodeName` while paused with a 409 conflict / validation error as originally proposed?

Author: Yes, a 409 conflict / validation error should return in this case.

Author: updated.
